src/tools/rbd_mirror/group_replayer/Replayer.cc

```diff
@@ -0,0 +1,1254 @@
+  if (itr != m_remote_group_snaps.end()) {
+    local_image_snap_specs.reserve(itr->snaps.size());
+    std::vector<cls::rbd::GroupImageStatus> local_images;
+    int r = local_group_image_list_by_id(&local_images);
+    if (r < 0) {
+      derr << "failed group image list: " << cpp_strerror(r) << dendl;
+      on_finish->complete(r);
+      return;
+    }
+    locker.unlock();
+    for (auto& image : local_images) {
+      std::unique_lock locker{m_lock};
+      bool image_snap_complete = false;
+      std::string image_header_oid = librbd::util::header_name(
+          image.spec.image_id);
+      ::SnapContext snapc;
+      int r = librbd::cls_client::get_snapcontext(&m_local_io_ctx,
+          image_header_oid, &snapc);
+      if (r < 0) {
+        derr << "get snap context failed: " << cpp_strerror(r) << dendl;
+        on_finish->complete(r);
+        return;
+      }
+
+      // stored in reverse order
+      for (auto snap_id : snapc.snaps) {
+        cls::rbd::SnapshotInfo snap_info;
+        r = librbd::cls_client::snapshot_get(&m_local_io_ctx, image_header_oid,
+            snap_id, &snap_info);
+        if (r < 0) {
+          derr << "failed getting snap info for snap id: " << snap_id
+            << ", : " << cpp_strerror(r) << dendl;
+          on_finish->complete(r);
+          return;
+        }
+
+        // extract { pool_id, snap_id, image_id }
+        auto ns = std::get_if<cls::rbd::ImageSnapshotNamespaceGroup>(
+            &snap_info.snapshot_namespace);
+        if (ns != nullptr && ns->group_snapshot_id == remote_group_snap_id) {
+          image_snap_complete = true;
+          cls::rbd::ImageSnapshotSpec snap_spec;
+          snap_spec.pool = image.spec.pool_id;
+          snap_spec.image_id = image.spec.image_id;
+          snap_spec.snap_id = snap_info.id;
+
+          local_image_snap_specs.push_back(snap_spec);
+        }
+      }
+      locker.unlock();
+      // In case if the image replayer is not yet updated with the limits, do it here
+      if (!image_snap_complete) {
+        set_image_replayer_limits(image.spec.image_id, &(*itr));
+      }
+      image_snap_complete = false; // reset
+    }
+  }
+  locker.lock();
+  if (itr->snaps.size() == local_image_snap_specs.size()) {
+    itl->snaps = local_image_snap_specs;
+    itl->state = cls::rbd::GROUP_SNAPSHOT_STATE_COMPLETE;
+  }
+  librados::ObjectWriteOperation op;
+  librbd::cls_client::group_snap_set(&op, *itl);
+
+  auto comp = create_rados_callback(
+      new LambdaContext([this, on_finish](int r) {
+        handle_regular_snapshot_complete(r, on_finish);
+      }));
+  int r = m_local_io_ctx.aio_operate(
+      librbd::util::group_header_name(m_local_group_id), comp, &op);
+  ceph_assert(r == 0);
+  comp->release();
+}
+
+template <typename I>
+void Replayer<I>::handle_regular_snapshot_complete(
+    int r, Context *on_finish) {
+  dout(10) << "r=" << r << dendl;
+  on_finish->complete(0);
```

Comment:
Should this be `on_finish->complete(r)` ? Why ignore error during `group_snap_set`?